### PR TITLE
Vulkan: fix draw array after draw rect bug

### LINKED ISSUE #169: Vulkan engine: draw array after draw rect doesn't work
`draw rect` command generates a new vertex buffer for the rectangle to given it.
When we implemented this, `GraphicsPipeline` just destroys the existing vertex buffer and creates new one.
As a result, `draw array` after `draw rect` looses its vertex buffer and has errors.

### PR DESCRIPTION
**Pull request by jaebaek**
`draw rect` command generates a new vertex buffer for its target
rectangle. Without this commit, GraphicsPipeline just destroys the
existing vertex buffer and creates new one. As a result, draw array
after draw rect looses its vertex buffer and has errors.

Fixes #169

### CODE REVIEW

File: src/vulkan/engine_vulkan.cc [hunk 1]
```diff
@@ -185,12 +189,14 @@ Result EngineVulkan::SetBuffer(BufferType type,
   if (!pipeline_)
     return Result("Vulkan::SetBuffer no Pipeline exists");
 
-  // TODO(jaebaek): Doublecheck those buffers are only for the graphics
-  //                pipeline.
   if (!pipeline_->IsGraphics())
     return Result("Vulkan::SetBuffer for Non-Graphics Pipeline");
 
-  pipeline_->AsGraphics()->SetBuffer(type, location, format, values);
+  if (!vertex_buffer_)
```
Comment (dj2) [inline src/vulkan/engine_vulkan.cc]: What happens if it's a kIndex buffer? Why do we create and set a vertex buffer in that case?

File: src/vulkan/graphics_pipeline.cc [hunk 2]
```diff
@@ -327,32 +328,38 @@ Result GraphicsPipeline::Initialize(uint32_t width,
...
-    return {};
+  if (type != BufferType::kVertex) {
+    return Result(
+        "GraphicsPipeline::SetVertexBuffer: non-vertex buffer type not "
+        "implemented");
+  }
 
-  if (!vertex_buffer_)
-    vertex_buffer_ = MakeUnique<VertexBuffer>(device_);
+  if (vertex_buffer == nullptr) {
```
Comment (dj2) [inline src/vulkan/graphics_pipeline.cc]: if (!vertex_buffer)

File: src/vulkan/graphics_pipeline.cc [hunk 3]
```diff
@@ -327,32 +328,38 @@ Result GraphicsPipeline::Initialize(uint32_t width,
...
-Result GraphicsPipeline::SetBuffer(BufferType type,
-                                   uint8_t location,
-                                   const Format& format,
-                                   const std::vector<Value>& values) {
+Result GraphicsPipeline::SetVertexBuffer(BufferType type,
+                                         uint8_t location,
+                                         const Format& format,
+                                         const std::vector<Value>& values,
+                                         VertexBuffer* vertex_buffer) {
   // TODO(jaebaek): Handle indices data.
```
Comment (dj2) [inline src/vulkan/graphics_pipeline.cc]: Why does SetVertexBuffer handle an Index buffer?

File: src/vulkan/graphics_pipeline.cc [hunk 4]
```diff
@@ -327,32 +328,38 @@ Result GraphicsPipeline::Initialize(uint32_t width,
...
     return r;
 
-  vertex_buffer_->SetData(location, format, values);
+  vertex_buffer->SetData(location, format, values);
   return {};
 }
 
-Result GraphicsPipeline::SendBufferDataIfNeeded() {
-  if (!vertex_buffer_)
+Result GraphicsPipeline::SendBufferDataIfNeeded(VertexBuffer* vertex_buffer) {
```
Comment (dj2) [inline src/vulkan/graphics_pipeline.cc]: Why does this take a VertexBuffer? Can't we send other types of buffer data as well?

File: src/vulkan/engine_vulkan.cc (hunk 1 above)
Comment (jaebaek) [reply src/vulkan/engine_vulkan.cc]: This is a mistake. Updated.

File: src/vulkan/graphics_pipeline.cc (hunk 3 above)
Comment (jaebaek) [reply src/vulkan/graphics_pipeline.cc]: removed

File: src/vulkan/graphics_pipeline.cc (hunk 2 above)
Comment (jaebaek) [reply src/vulkan/graphics_pipeline.cc]: done

File: src/vulkan/graphics_pipeline.cc (hunk 4 above)
Comment (jaebaek) [reply src/vulkan/graphics_pipeline.cc]: Changed it as `SendVertexBufferDataIfNeeded`

**Pull request merged**